src/Rally.ts

```diff
@@ -102,90 +102,113 @@ export class Rally {
-            this.pause();
-          }
-        } else {
-          const userStudiesDoc = await getDoc(doc(this._db, "users", uid, "studies", this._options.studyId));
-          // TODO do runtime validation of this document
-          if (userStudiesDoc && !userStudiesDoc.exists()) {
-            // This document is created by the site and may not exist yet.
-            console.warn("Rally.onSnapshot - userStudies document does not exist yet");
-            return;
-          }
+  // Monitors user record at: extensionUsers/<uid>
+  private monitorUserRecord(uid: string) {
+    // This contains the Rally ID, need to call the Rally state change callback with it.
+    onSnapshot(doc(this._db, "extensionUsers", uid), extensionUserDoc => {
+      if (!extensionUserDoc.exists()) {
+        // User record is either not present or is somehow deleted
+        // hence change the state to be logged out
+        this._signedIn = false;
 
-          const data = userStudiesDoc.data();
+        throw new Error("Rally onSnapshot - extensionUser document does not exist");
+      }
 
-          if (data.enrolled && this._state !== RunStates.Running) {
-            this.resume();
-          }
-        }
+      // https://datatracker.ietf.org/doc/html/rfc4122#section-4.1.7
+      const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
 
-        if (data.studyEnded === true) {
-          if (this._state !== RunStates.Ended) {
-            this.end();
-          }
+      const data = extensionUserDoc.data();
+
+      if (data && data.rallyId) {
+        if (data.rallyId.match(uuidRegex)) {
+          // Stored Rally ID looks fine, cache it and call the Rally state change callback with it.
+          this._rallyId = data.rallyId;
+        } else {
+          // Do not loop or destroy data if the stored Rally ID is invalid, bail out instead.
+          throw new Error(`Stored Rally ID is not a valid UUID: ${data.rallyId}`);
         }
-      });
+      }
+    });
+  }
 
-      onSnapshot(doc(this._db, "users", uid, "studies", this._options.studyId), async userStudiesDoc => {
-        if (!userStudiesDoc.exists()) {
+  // Monitor user's current study record at: users/<uid>/studies/<studyId> and pause/resume data collection.
+  private monitorCurrentStudyRecordForUser(uid: string) {
+    onSnapshot(doc(this._db, "users", uid, "studies", this._options.studyId), async userStudiesDoc => {
+      if (!userStudiesDoc.exists()) {
+        // This document is created by the site and may not exist yet.
+        console.warn("Rally.onSnapshot - userStudies document does not exist");
+        return;
+      }
+
+      const data = userStudiesDoc.data();
+      if (data.enrolled) {
+        this.resume();
+      } else {
+        this.pause();
+      }
+    });
+  }
+
+  // Monitor current study record at: studies/<studyId> and pause/resume/end data collection.
+  private monitorCurrentStudyRecord(uid: string) {
```

Comment:
Seems semantically odd to pass user id to a function monitoring a study record, and not pass the study id. Maybe it makes sense to make `uid` into a private member variable as well? Or maybe just change it to `userId` so it's clear.

Reply:
Used closure to address this now.